libs/protocols/bmp180.c

```diff
@@ -0,0 +1,402 @@
+					json_append_member(code, "temperature", json_mknumber(((double) temp / 10) + temp_offset, 1)); // in deg C
+					json_append_member(code, "pressure", json_mknumber(((double) pressure / 100) + pressure_offset, 1)); // in hPa
+
+					json_append_member(bmp180->message, "message", code);
+					json_append_member(bmp180->message, "origin", json_mkstring("receiver"));
+					json_append_member(bmp180->message, "protocol", json_mkstring(bmp180->id));
+
+					pilight.broadcast(bmp180->id, bmp180->message);
+					json_delete(bmp180->message);
+					bmp180->message = NULL;
+				} else {
+					logprintf(LOG_DEBUG, "error connecting to bmp180");
+					logprintf(LOG_DEBUG, "(probably i2c bus error from wiringXI2CSetup)");
+					logprintf(LOG_DEBUG, "(maybe wrong id? use i2cdetect to find out)");
+					protocol_thread_wait(node, 1, &nrloops);
+				}
+			}
+			pthread_mutex_unlock(&bmp180lock);
+#endif
+		}
+	}
+
+	if (bmp180data->id) {
+		for (y = 0; y < bmp180data->nrid; y++) {
+			sfree((void *) &bmp180data->id[y]);
+		}
+		sfree((void *) &bmp180data->id);
+	}
+	if (bmp180data->fd) {
+		for (y = 0; y < bmp180data->nrid; y++) {
+			if (bmp180data->fd[y] > 0) {
+				close(bmp180data->fd[y]);
+			}
+		}
+		sfree((void *) &bmp180data->fd);
+	}
+	sfree((void *) &bmp180data);
+	bmp180_threads--;
+
+	return (void *) NULL;
+}
+
+struct threadqueue_t *bmp180InitDev(JsonNode *jdevice) {
+	bmp180_loop = 1;
+	wiringXSetup();
+	char *output = json_stringify(jdevice, NULL);
+	JsonNode *json = json_decode(output);
+	sfree((void *) &output);
+
+	struct protocol_threads_t *node = protocol_thread_init(bmp180, json);
+	return threads_register("bmp180", &bmp180Parse, (void *) node, 0);
+}
+
+static void bmp180ThreadGC(void) {
+	bmp180_loop = 0;
+	protocol_thread_stop(bmp180);
+	while (bmp180_threads > 0) {
+		usleep(10);
+	}
+	protocol_thread_free(bmp180);
+}
+
+static void bmp180PrintHelp(void) {
+	printf("\t -i --id=id\t\t\tcontrol a device with this id\n");
+	printf("\t -o --oversampling=oversampling\tset the oversampling mode (0, 1, 2, 3)\n");
+	printf("\t -p --pressure=pressure\t\tset the barometric pressure in hPa\n");
+	printf("\t -t --temperature=temperature\tset the temperature in deg C\n");
+}
+
+#ifndef MODULE
+__attribute__((weak))
+#endif
+void bmp180Init(void) {
+	pthread_mutexattr_init(&bmp180attr);
+	pthread_mutexattr_settype(&bmp180attr, PTHREAD_MUTEX_RECURSIVE);
+	pthread_mutex_init(&bmp180lock, &bmp180attr);
+
+	protocol_register(&bmp180);
+	protocol_set_id(bmp180, "bmp180");
+	protocol_device_add(bmp180, "bmp180", "I2C Barometric Pressure and Temperature Sensor");
+	protocol_device_add(bmp180, "bmp085", "I2C Barometric Pressure and Temperature Sensor");
+	bmp180->devtype = WEATHER;
+	bmp180->hwtype = SENSOR;
+
+	options_add(&bmp180->options, 'i', "id", OPTION_HAS_VALUE, DEVICES_ID, JSON_STRING, NULL, "0x[0-9a-f]{2}");
+	options_add(&bmp180->options, 'o', "oversampling", OPTION_HAS_VALUE, DEVICES_SETTING, JSON_NUMBER, (void *) 1,
```

Comment:
Option `oversampling` sets the BMP180 hardware accuracy mode, selected by driver software via the variable `oversampling_setting`. This is only relevant for pressure readings (and not for temperature readings).

Mode | Parameter oversampling_setting | Internal no. of samples | Conversion time pressure max. [ms]
-----|--------------------------------|-------------------------|-----------------------------------
ultra low power | 0 | 1 | 4.5
standard | 1 | 2 | 7.5
high resolution | 2 | 4 | 13.5 
ultra high resolution | 3 | 8 | 25.5